src/units/abilities.cpp

```diff
@@ -1288,6 +1288,18 @@ int attack_type::modified_damage() const
 	return damage_value;
 }
 
+unit_alignments::type attack_type::alignment() const
+{
+	// pick attack alignment or fall back to unit alignment
+	return (unit_alignments::get_enum(alignment_str_).value_or(self_ ? self_->alignment() : unit_alignments::type::neutral));
+}
+
+std::string attack_type::alignment_str() const
+{
+	//if self_return the alignment select by alignment() else return the original string.
+	return (self_ ? unit_alignments::get_string(alignment()): alignment_str_);
```

Comment:
Can you explain in detail how to see the problem why this should be needed? Perhaps open an issue about it so other people can take a look.

For this PR I suggest to remove this function.

Reply:
i don't want attack alignment return neutral when attribute empty in lua code where i can't put unit data

Reply:
In the impl_unit_attack_get() function in the lua unit attacks.cpp file the unit information is not present despite all my efforts to solve this problem, which means that if I return a simple reverse transcription of attack alignment in string but alignment_ is empty or badly transcribed then it will be interpreted as a neutral alignment to avoid this while waiting to find a more appropriate solution, in case of absence of self_ and therefore of unit alignment, the string should simply alignment_str_ without selection, so if it is empty, it is empty and not neutral (empty means no alignment and not neutral alignment), when self_ present could return unit alignment if weapon alignment no valid.

Reply:
We can't be doing weird workarounds for the fact that the attack does not have `self_` when you access it from Lua. This is not a bug, exactly. It is a missing feature. What we need is a way for Lua to set the `self_` on an attack; but if the attack for whatever reason doesn't have it set, then it should return the proper data that you'd expect in such a case (probably the base data).

Reply:
i all try for resolve problem of uint data in this table without succes, someone can help me?

Reply:
for me the proper data in absence of self_ in lua this is the base data (alignment_str_) without modification and i must let the function like it is or use a base_alignment() function for return directly alignment_str_

Reply:
> What I recall from what was discussed on discord is that this isn't possible to do.

what it isn't possible to do?

Reply:
You can check the IRC logs from the 19th, but:
> I don't think impl_unit_attack_get has a way to access the unit who owns the attack. There might not even be one. It could be an attack belonging to unit type, or a standalone attack.

Reply:
Btw, as far as I'm concerned I'm fine with merging this (with a sane interface not in the current state) even if usage from lua does not work right. The fix needs to be done in the lua integration and I'm fairly certain the fix is not just needed for this. (IMO there is a good chance that #9232 and #9206 for example are related.)

Reply:
whaat issue has the interface? For lua i don't know how fix this.

Reply:
The interface should not have functions that do special logic just for lua. As I'm sure if said multiple times by now IMO there should be a single function that returns the alignment and that's it.

Reply:
i changed alignment_str() code for handle lua case, i can't doing best.

Reply:
I know, it one alternative is to remove lua code.

Reply:
@soliton- if i removed attribute of lua code,in wait of fix o lua integration, will you merge this PR?

Reply:
If you mean not allowing lua to access the alignment attribute for now then I guess that should be fine as well.

Reply:
removed from lua.

Reply:
@soliton- @stevecotton resolved issue with https://github.com/wesnoth/wesnoth/pull/8851/commits/f5fd70d16800238bb8d5374fccdef46f71305da4

Reply:
Ok, missed this before merging. Duplicating the existing alignments there does not look good. Oh well...

Reply:
This is the only solution that works in Lua and I can't put self_ in attack_type() in all files where it is called because it is not always called with an attack_type variable, this hybrid solution is the only one that works and I need that when attack_type() does not have self_ and alignment is empty or invalid that alignment_str_ is empty and not replaced by a neutral alignment so that alignment() can return the unit alignment instead of attack_type()

Reply:
My point was that you duplicated the alignment strings instead of using `unit_alignments::get_enum` or so to check.